src/main/java/bio/terra/stairway/FlightMap.java

```diff
@@ -77,33 +102,78 @@ public void makeImmutable() {
    * @throws ClassCastException if found, not castable to the requested type
    */
   public <T> T get(String key, Class<T> type) {
-    Object value = map.get(key);
+    String value = map.get(key);
     if (value == null) {
       return null;
     }
 
-    if (type.isInstance(value)) {
-      return type.cast(value);
+    try {
+      return getObjectMapper().readValue(value, type);
+    } catch (final JsonProcessingException ex) {
+      throw new ClassCastException(
+          "Found value '" + value + "' is not an instance of type " + type.getName());
     }
-    throw new ClassCastException(
-        "Found value '" + value.toString() + "' is not an instance of type " + type.getName());
   }
 
+  /**
+   * Gets the raw (serialized) data stored for an Object in the hash map.
+   *
+   * @param key Key to lookup the data for in the map.
+   * @return String mapped to key, or null if key does not exist.
+   */
+  public String getRaw(String key) {
+    return map.get(key);
+  }
+
+  /**
+   * Serialize an Object and place it into the map at a given key.
+   *
+   * @param key Key to store the object under.
+   * @param value Object to serialize and store.
+   * @throws JsonConversionException if object could not be serialized.
+   */
   public void put(String key, Object value) {
+    try {
+      map.put(key, getObjectMapper().writeValueAsString(value));
+    } catch (final JsonProcessingException ex) {
+      throw new JsonConversionException("Failed to convert object to json string", ex);
+    }
+  }
+
+  /**
+   * Stores a raw (serialized) data value for a given key.
+   *
+   * @param key Key to store the data under.
+   * @param value Data to store.
+   */
+  public void putRaw(String key, String value) {
     map.put(key, value);
   }
 
   public String toJson() {
     try {
-      return getObjectMapper().writeValueAsString(map);
-    } catch (JsonProcessingException ex) {
+      // Serializing as Map<String, RawJsonValue> will produce a format compatible with older code's
+      // Map<String, Object> representation.
+      Map<String, RawJsonValue> temporaryMap = new HashMap<>();
+      for (Map.Entry<String, String> entry : map.entrySet()) {
+        temporaryMap.put(entry.getKey(), RawJsonValue.create(entry.getValue()));
```

Comment:
Why not just perform the serialization of the object here?
It looks to me like going through the RawJsonType and custom serializer stuff does not get a different result, but spreads the logic out a lot making it harder to follow.

Reply:
To this question and @wchamber question above, I was going to some lengths to use raw JSON semantics to enforce the restriction that we only ever ser/des **`Objects`** at get/put time, and traffic in JSON everywhere else.

If we are OK with relaxing that I can see if this can be simplified by deserializing and re-serializing into a `Map<String, Object>` and getting the same result.

The spreading out of the logic is due to not wanting to re-register the same serialization hooks with the static global ObjectMapper on every call, thus making it available to the global static instance creation.

Reply:
Map<String, String> change needs to wait for the database changes to propagate due to the need to write backward compatible JSON maps.  Reverted most of FlightMap.